Gems/PhysX/Code/Editor/Source/ComponentModes/Joints/JointsComponentMode.h

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+#pragma once
+
+#include <AzCore/EBus/Event.h>
+#include <AzCore/Math/Uuid.h>
+#include <AzCore/Memory/Memory.h>
+#include <AzCore/std/containers/unordered_map.h>
+#include <AzCore/std/containers/vector.h>
+#include <AzCore/std/smart_ptr/unique_ptr.h>
+#include <AzToolsFramework/ComponentMode/EditorBaseComponentMode.h>
+#include <AzToolsFramework/ViewportUi/ViewportUiRequestBus.h>
+
+#include <Editor/Source/ComponentModes/Joints/JointsComponentModeCommon.h>
+#include <Editor/Source/ComponentModes/PhysXSubComponentModeBase.h>
+
+namespace AZ
+{
+    class EntityComponentIdPair;
+}
+
+namespace PhysX
+{
+    class JointsComponentMode final
+        : public AzToolsFramework::ComponentModeFramework::EditorBaseComponentMode
+    {
+    public:
+        AZ_CLASS_ALLOCATOR_DECL;
+
+        JointsComponentMode(const AZ::EntityComponentIdPair& entityComponentIdPair, AZ::Uuid componentType);
+        ~JointsComponentMode();
+
+        // EditorBaseComponentMode ...
+        void Refresh() override;
+        AZStd::vector<AzToolsFramework::ActionOverride> PopulateActionsImpl() override;
+        AZStd::vector<AzToolsFramework::ViewportUi::ClusterId> PopulateViewportUiImpl() override;
+    private:
+        enum class ClusterGroups
+        {
+            Group1 = 0,
+            Group2,
+            Group3,
+
+            NumGroups
+        };
+
+        struct ButtonData
+        {
+            AzToolsFramework::ViewportUi::ClusterId m_clusterId;
+            AzToolsFramework::ViewportUi::ButtonId m_buttonId;
+        };
+
+        void SetCurrentMode(JointsComponentModeCommon::SubComponentModes::ModeType newMode, ButtonData& buttonData);
+        bool HandleMouseInteraction(const AzToolsFramework::ViewportInteraction::MouseInteractionEvent& mouseInteraction) override;
+        void SetupSubModes(const AZ::EntityComponentIdPair& entityComponentIdPair);
+        void ResetCurrentMode();
+        void TeardownSubModes();
+
+        AzToolsFramework::ViewportUi::ClusterId GetClusterId(ClusterGroups group);
+
+        AZStd::vector<AzToolsFramework::ViewportUi::ClusterId> m_modeSelectionClusterIds;
+        AZStd::unordered_map<JointsComponentModeCommon::SubComponentModes::ModeType, ButtonData> m_buttonData;
+
+        AZStd::vector<AZ::Event<AzToolsFramework::ViewportUi::ButtonId>::Handler> m_modeSelectionHandlers;
+        ButtonData m_activeButton;
+
+        JointsComponentModeCommon::SubComponentModes::ModeType m_subMode = JointsComponentModeCommon::SubComponentModes::ModeType::Translation;
+        AZStd::unordered_map<JointsComponentModeCommon::SubComponentModes::ModeType, AZStd::unique_ptr<PhysXSubComponentModeBase>> m_subModes;
+    };
+}
```

Comment:
uber nit: closing namespace (could run clang-format on the new files too which will do all this and more for you)